reports/src/main/java/com/ivy/reports/ReportFlowViewModel.kt

```diff
@@ -0,0 +1,413 @@
+            AmountType.MIN -> filter.copy(minAmount = amt)
+            AmountType.MAX -> filter.copy(maxAmount = amt)
+        }
+
+        updateFilter(newFilter = newFilter)
+    }
+
+    private fun updateIncludeExcludeKeywords(
+        filter: FilterState,
+        keyword: String,
+        keywordsType: KeywordsType,
+        add: Boolean
+    ) {
+        val trimmedKeyword = keyword.trim()
+
+        val existingKeywords = when (keywordsType) {
+            KeywordsType.INCLUDE -> filter.includeKeywords.data
+            KeywordsType.EXCLUDE -> filter.excludeKeywords.data
+        }
+
+        val updatedKeywords = if (add && trimmedKeyword !in existingKeywords)
+            existingKeywords.addOrRemove(add = true, trimmedKeyword)
+        else if (!add)
+            existingKeywords.addOrRemove(add = false, trimmedKeyword)
+        else
+            existingKeywords
+
+
+        val newFilter = when (keywordsType) {
+            KeywordsType.INCLUDE -> filter.copy(includeKeywords = updatedKeywords.toImmutableItem())
+            KeywordsType.EXCLUDE -> filter.copy(excludeKeywords = updatedKeywords.toImmutableItem())
+        }
+
+        updateFilter(newFilter = newFilter)
+    }
+
+    private fun updatePlannedPaymentSelection(
+        filter: FilterState,
+        type: PlannedPaymentTypes,
+        add: Boolean
+    ) {
+        val selectPlannedPayments =
+            filter.selectedPlannedPayments.data.addOrRemove(add = add, item = type)
+                .toImmutableItem()
+
+        updateFilter(newFilter = filter.copy(selectedPlannedPayments = selectPlannedPayments))
+    }
+
+    private fun onFilterClear(
+        filter: FilterState,
+        event: ReportFilterEvent.Clear
+    ) {
+        val newFilter = when (event) {
+            is ReportFilterEvent.Clear.Filter -> emptyFilterState()
+            ReportFilterEvent.Clear.Accounts -> filter.copy(selectedAcc = emptyList<Account>().toImmutableItem())
+            ReportFilterEvent.Clear.Categories -> filter.copy(selectedCat = emptyList<Category>().toImmutableItem())
+        }
+        updateFilter(newFilter)
+    }
+
+
+    private fun onSelectAll(
+        filter: FilterState,
+        event: ReportFilterEvent.SelectAll
+    ) {
+        val newFilter = when (event) {
+            is ReportFilterEvent.SelectAll.Accounts -> filter.copy(
+                selectedAcc = filter.allAccounts.data.toList().toImmutableItem()
+            )
+            ReportFilterEvent.SelectAll.Categories -> filter.copy(
+                selectedCat = filter.allCategories.data.toList().toImmutableItem()
+            )
+        }
+
+        updateFilter(newFilter = newFilter)
+    }
+
+    private fun updateFilter(newFilter: FilterState) {
+        filter.value = newFilter
+    }
+
+    private fun initialiseData() {
+        combine(accountsFLow(), categoryFlow()) { a, c ->
+            Pair(a, c)
+        }.onEach { pair ->
+            filter.update {
```

Comment:
That's a workaround

Reply:
Any alternatives?

Reply:
Move allAccounts out of `filter` and filter mutable state can contain only selected accs and categories? Would that handle all cases? @Vishwa-Raghavendra